src/modules/job-info/lookup.c

```diff
@@ -102,58 +100,137 @@ static flux_future_t *lookup_key (struct lookup_ctx *l, const char *key,
         goto error;
     }
 
-    if (flux_future_then (f, -1, c, l) < 0) {
+    if (c) {
+        if (flux_future_then (f, -1, c, l) < 0) {
+            flux_log_error (l->ctx->h, "%s: flux_future_then", __FUNCTION__);
+            goto error;
+        }
+    }
+
+    return f;
+
+error:
+    save_errno = errno;
+    flux_future_destroy (f);
+    errno = save_errno;
+    return NULL;
+}
+
+static flux_future_t *lookup_keys (struct lookup_ctx *l, flux_continuation_f c)
+{
+    flux_future_t *fall = NULL;
+    flux_future_t *f = NULL;
+    size_t index;
+    json_t *key;
+    int save_errno;
+
+    if (!(fall = flux_future_wait_all_create ())) {
+        flux_log_error (l->ctx->h, "%s: flux_wait_all_create", __FUNCTION__);
+        goto error;
+    }
+    flux_future_set_flux (fall, l->ctx->h);
+
+    json_array_foreach(l->keys, index, key) {
+        const char *keystr;
+        if (!(keystr = json_string_value (key))) {
+            errno = EINVAL;
+            goto error;
+        }
+        if (!(f = lookup_key (l, keystr, NULL)))
+            goto error;
+        if (flux_future_push (fall, keystr, f) < 0) {
+            flux_log_error (l->ctx->h, "%s: flux_future_push", __FUNCTION__);
+            goto error;
+        }
+    }
+
+    if (flux_future_then (fall, -1, c, l) < 0) {
```

Comment:
If this fails, `f` is freed but it was already added to `fall` so double free?

Note that `flux_future_destroy()` doesn't clobber errno so it's not strictly necessary to save/restore (which is convenient to know if you have to do the free in the `flux_future_push()` error case rather than in `error:`.